samza-core/src/main/java/org/apache/samza/storage/blobstore/util/BlobStoreUtil.java

```diff
@@ -670,12 +676,18 @@ private CompletableFuture<Void> removeTTL(DirIndex dirIndex, Metadata metadata)
               metadata.getJobName(), metadata.getJobId(), metadata.getTaskName(), metadata.getStoreName());
       List<FileBlob> fileBlobs = file.getBlobs();
       for (FileBlob fileBlob : fileBlobs) {
-        String opname = "removeTTL for fileBlob: " + file.getFileName() + " with blobId: {}" + fileBlob.getBlobId();
-        Supplier<CompletionStage<Void>> ttlRemovalAction = () ->
-            blobStoreManager.removeTTL(fileBlob.getBlobId(), requestMetadata).toCompletableFuture();
-        CompletableFuture<Void> ttlRemovalFuture =
-            FutureUtil.executeAsyncWithRetries(opname, ttlRemovalAction, isCauseNonRetriable(), executor, retryPolicyConfig);
-        updateTTLsFuture.add(ttlRemovalFuture);
+        if (ttlRemovedBlobIdsCache.getIfPresent(fileBlob.getBlobId()) == null) {
+          String opname = "removeTTL for fileBlob: " + file.getFileName() + " with blobId: {}" + fileBlob.getBlobId();
+          Supplier<CompletionStage<Void>> ttlRemovalAction = () ->
+              blobStoreManager.removeTTL(fileBlob.getBlobId(), requestMetadata).toCompletableFuture()
+                  .thenRun(() -> ttlRemovedBlobIdsCache.put(fileBlob.getBlobId(), new Object()));
```

Comment:
Lets also store the time this was removed for debugging purposes and add it in the log below

Reply:
ACK. Will add that to the cache and to debug lines. Thanks.